src/db.c

```diff
@@ -483,12 +499,23 @@ int dbGenericDelete(redisDb *db, robj *key, int async, int flags) {
         /* We want to try to unblock any module clients or clients using a blocking XREADGROUP */
         signalDeletedKeyAsReady(db,key,val->type);
         /* We should call decr before freeObjAsync. If not, the refcount may be
-         * greater than 1, so freeObjAsync doesn't work */
-        decrRefCount(val);
+         * greater than 1, so freeObjAsync doesn't work .
+         *
+         * With value-embedding, if the val undergoes dbUnshareStringValue operation, the provided object
+         * would replaced from the dictionary and a new entry would be created. Hence, the refcount might
+         * have been reduced already as well as the OBJ_STATE_PROTECTED would have been removed. Hence,
+         * we make the val object refcount/state ideal for dictTwoPhaseUnlinkFree to clean the val */
```

Comment:
I thought reading this a second time would make more sense, but this code really seems like a hack. It seems like the logic should really be more in dbUnshareStringValue() than in here.

Reply:
Yeah, I also agree this is a hack. Need to think through how to handle this scenario more cleanly.